google-analytics-dashboard.html

```diff
@@ -32,66 +32,123 @@
-        }.bind(this))
+      authorized: {
+        type: Boolean,
+        value: false,
+        reflectToAttribute: true
+      },
 
-        this.updateChildren();
+      errorMessage: {
+        type: String,
+        value: ''
       },
 
+      hasError: {
+        type: Boolean,
+        computed: 'computeHasError(errorMessage)'
+      }
+    },
+
+    listeners: {
+      'analytics-dashboard-control-change': 'queryUpdated',
+      'analytics-query-error': 'handleQueryError'
+    },
+    ready: function() {
       /**
-       * The `updateChildren` method updates each of this dashboards
-       * `<google-analytics-chart>` element with its current query value.
+       * The `query` attribute represents the internal query object of this
+       * dashboard. It is updated when control elements fire the
+       * `analytics-dashboard-control-change` event and pass along query data.
        *
-       * @method updateChildren
+       * @attribute query
+       * @type object
        */
-      updateChildren: function() {
-        var charts = this.querySelectorAll('google-analytics-chart');
-        for (var i = 0, chart; chart = charts[i]; i++) {
-          Object.keys(this.query).forEach(function(key) {
-            chart[key] = this.query[key];
-          }.bind(this));
-        }
-      }
 
-    });
-
-  </script>
-
-</polymer-element>
+     this.updateChildren();
+    },
+
+    /**
+     * The `queryUpdated` method is the callback for whenever the
+     * `analytics-dashboard-control-change` event is fired. It updates the
+     * query attribute, which is then sent to child charts.
+     *
+     * @method queryUpdated
+     * @param {CustomEvent} event - The event with the query data.
+     */
+    queryUpdated: function(event) {
+      // Update `this.query` with the passed event data.
+      Object.keys(event.detail).forEach(function(key) {
+        this.query[key] = event.detail[key];
+      }.bind(this))
+
+      this.updateChildren();
+    },
+
+    /**
+     * The `updateChildren` method updates each of this dashboards
+     * `<google-analytics-chart>` element with its current query value.
+     *
+     * @method updateChildren
+     */
+    updateChildren: function() {
+      if (!this.authorized)
+        return;
+      var charts = Polymer.dom(this).querySelectorAll('google-analytics-chart');
+      for (var i = 0, chart; chart = charts[i]; i++) {
+        Object.keys(this.query).forEach(function(key) {
+          chart[key] = this.query[key];
+        }.bind(this));
+      }
+    },
+    signedIn: function() {
+      this.authorized = true;
+      this.updateChildren();
+    },
+    signedOut: function() {
+      this.authorized = false;
+    },
+    handleQueryError: function(ev) {
+      this.errorMessage = ev.detail;
+      // Invalid credentials after a multi-hour session have been observed in the wild.
+      this.errorMessage = "Error loading data: " + ev.detail.message + ". Your session might have expired.";
```

Comment:
I think this message is helpful in some cases (well, the case where your session expired), but it's unhelpful for all other error messages. For example I just go this:

> Error loading data: Quota Error: User Rate Limit Exceeded.. Your session might have expired.

Also, messages returned from the API seem to have periods at the end, so adding the extra period makes the message appear to have a partial ellipsis.